{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#!usr/bin/env python\n",
    "\n",
    "######################################## Format argparse ##########################################################\n",
    "# 1. Write script parameters to take in the four files as well as the index.tsv file and a min quality score cut off.\n",
    "\n",
    "import argparse\n",
    "import numpy as np\n",
    "\n",
    "parser = argparse.ArgumentParser(description=\"Quality Index Swapping\")\n",
    "parser.add_argument(\"-f1\", \"--filepath1\", help=\"R1 file\", required=True, type=str)\n",
    "parser.add_argument(\"-f2\", \"--filepath2\", help=\"R2 file\", required=True, type=str)\n",
    "parser.add_argument(\"-f3\", \"--filepath3\", help=\"R3 file\", required=True, type=str)\n",
    "parser.add_argument(\"-f4\", \"--filepath4\", help=\"R4 file\", required=True, type=str)\n",
    "parser.add_argument(\"-f5\", \"--filepath5\", help=\"Indexes.txt file\", required=True, type=str)\n",
    "parser.add_argument(\"-min_qscore\", \"--min_cutoff\", help=\"Min quality score cutoff\", required=True, type=int)\n",
    "\n",
    "\n",
    "args = parser.parse_args() #sets arguments as call-able \n",
    "f1 = args.filepath1\n",
    "f2 = args.filepath2\n",
    "f3 = args.filepath3\n",
    "f4 = args.filepath4\n",
    "f5 = args.filepath5\n",
    "min_qscore = args.min_cutoff\n",
    "\n",
    "\n",
    "def convert_phred(letter):\n",
    "    n = ord(letter) - 33\n",
    "    return n\n",
    "\n",
    "\n",
    "######################## Build dictionary containing every 4 nucleotide seq combination ############################\n",
    "\n",
    "# Every possible combo of 2, 8 base long indexes as key and value = counter(increases by 1 every time seq is \n",
    "# uncovered in index reads R2, R3)\n",
    "N_Idx_Dic = {}\n",
    "\n",
    "index_values={} #dictionary to hold all combinations of index values from said FASTQ/Index files\n",
    "list_of_index_reads=[]#list holding index reads from said index file\n",
    "un_matched_index = {} #list that has index reads \n",
    "\n",
    "\n",
    "\n",
    "            \n",
    "with open(f5) as tsv_fh: #open tsv file, set to filehandle (fh)\n",
    "    for line in tsv_fh:\n",
    "        values_per_column=line.strip().split(\"\\t\") #strip and split by tab for each value in both columns\n",
    "        index_reads=(values_per_column[4]) #grab index reads\n",
    "        list_of_index_reads.append(index_reads) #put into list_of_index_reads array\n",
    "    #print(list_of_index_reads)\n",
    "\n",
    "#generate all possible combonations of 2 index values and store in dict with value of 0\n",
    "for line_in_file in list_of_index_reads:\n",
    "    for indexes in list_of_index_reads:\n",
    "        if line_in_file == indexes:\n",
    "            index_values[line_in_file + \"_\" + indexes]=0#set indexes to 0 [index1,index2:0]\n",
    "        else:\n",
    "            un_matched_index[line_in_file + \"_\" + indexes] = 0  #Generate underscore between each line for indexes in dic        \n",
    "\n",
    "\n",
    "############################################### Setting variables/read in files ##########################################################\n",
    "\n",
    "head1 = \"\" #setting header variable for each corresponding first line in each read of the FASTQ file\n",
    "head2 = \"\"\n",
    "head3 = \"\"\n",
    "head4 = \"\"\n",
    "\n",
    "seq1 = \"\" #setting sequence variable for each corresponding second line in each read of the FASTQ file\n",
    "seq2 = \"\"\n",
    "seq3 = \"\"\n",
    "seq4 = \"\"\n",
    "\n",
    "plus_symbol1 = \"\" #setting (+) symbol to variable for each corresponding third line in each read of the FASTQ file\n",
    "plus_symbol2 = \"\"\n",
    "plus_symbol3 = \"\"\n",
    "plus_symbol4 = \"\"\n",
    "\n",
    "q_score1 = \"\" #setting qscore variable for each corresponding fourth line in each read of the FASTQ file\n",
    "q_score2 = \"\"\n",
    "q_score3 = \"\"\n",
    "q_score4 = \"\"\n",
    "\n",
    "\n",
    "\n",
    "############Setting global vairables for each said counter below############\n",
    "N_counter = 0\n",
    "matched_counter = 0\n",
    "idx_hop_counter = 0\n",
    "sequencing_error_counter = 0\n",
    "retained_values_counter = 0\n",
    "discarded_values_counter = 0\n",
    "total_counter = 0\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "def rc(dna):\n",
    "    sequence_dictionary = {'A':'T','T':'A','G':'C','C':'G', 'N':'N'}\n",
    "    return \"\".join([sequence_dictionary[base] for base in reversed(dna)])\n",
    " \n",
    "\n",
    "    \n",
    "    \n",
    "    \n",
    "##################################### Assigning each line in FASTQ file to a variable ######################################\n",
    "    \n",
    "    \n",
    "with open(f1, \"rt\") as fh_R1, open(f2, \"rt\") as fh_R2, open(f3) as fh_R3, open(f4) as fh_R4:\n",
    "    for line in fh_R1:\n",
    "        line = line.strip(\"\\n\")\n",
    "        head1 = line.strip(\"\\n\")\n",
    "        seq1 = fh_R1.readline().strip(\"\\n\")\n",
    "        plus_symbol1 = fh_R1.readline().strip(\"\\n\")\n",
    "        q_score1 = fh_R1.readline().strip(\"\\n\")\n",
    "\n",
    "        head2 = fh_R2.readline().strip(\"\\n\")\n",
    "        seq2 = fh_R2.readline().strip(\"\\n\")\n",
    "        plus_symbol2 = fh_R2.readline().strip(\"\\n\")\n",
    "        q_score2 = fh_R2.readline().strip(\"\\n\")\n",
    "\n",
    "        head3 = fh_R3.readline().strip(\"\\n\")\n",
    "        seq3 = rc(fh_R3.readline().strip(\"\\n\"))\n",
    "        plus_symbol3 = fh_R3.readline().strip(\"\\n\")\n",
    "        q_score3 = fh_R3.readline().strip(\"\\n\")\n",
    "\n",
    "        head4 = fh_R4.readline().strip(\"\\n\")\n",
    "        seq4 = fh_R4.readline().strip(\"\\n\")\n",
    "        plus_symbol4 = fh_R4.readline().strip(\"\\n\")\n",
    "        q_score4 = fh_R4.readline().strip(\"\\n\")\n",
    "\n",
    "        value_of_index = seq2 + \"_\" + seq3\n",
    "\n",
    "        total_counter += 1\n",
    "        \n",
    "        if total_counter%500000000:\n",
    "            print(\"On line: \", total_counter, flush = True)\n",
    "\n",
    "        raw_score = []\n",
    "        for base in range(len(q_score2)):        \n",
    "            raw_score.append(convert_phred(q_score2[base]))\n",
    "            raw_score.append(convert_phred(q_score3[base]))\n",
    "        \n",
    "        if all(items >= min_qscore for items in raw_score):  \n",
    "            retained_values_counter+=1 \n",
    "            if value_of_index in index_values:           \n",
    "                index_values[value_of_index] += 1\n",
    "                matched_counter += 1\n",
    "            \n",
    "            if value_of_index in un_matched_index:    \n",
    "                un_matched_index[value_of_index] += 1\n",
    "                idx_hop_counter += 1\n",
    "            \n",
    "            if \"N\" in value_of_index:\n",
    "                if value_of_index in N_Idx_Dic:\n",
    "                    N_Idx_Dic[value_of_index]+=1\n",
    "                    N_counter += 1\n",
    "                else:\n",
    "                    N_Idx_Dic[value_of_index]=1\n",
    "                    N_counter += 1\n",
    "            \n",
    "            if value_of_index not in index_values and value_of_index not in un_matched_index and value_of_index not in N_Idx_Dic:\n",
    "                sequencing_error_counter += 1      \n",
    "    \n",
    "        else:\n",
    "            discarded_values_counter += 1 \n",
    "            \n",
    "##################################### Outputting stats and dictionary file ############################################\n",
    "            \n",
    "            \n",
    "with open(\"1294_S1_L008_R\"+ \"_Output_table_cov_0\", 'w') as output1:\n",
    "    j = 0\n",
    "    m = 0   \n",
    "    k = 0\n",
    "    output1.write(\"\\n\" + \"Number of Undetermined N Reads: \" + str(N_counter))\n",
    "    output1.write(\"\\n\" + \"Number of Expected Paired Index Reads : \" + str(matched_counter))\n",
    "    output1.write(\"\\n\" + \"Number of Index Swapped Reads: \" + str(idx_op_counter))\n",
    "    output1.write(\"\\n\" + \"Number of Sequencing Error Reads: \" + str(sequencing_error_counter))\n",
    "    output1.write(\"\\n\" + \"Number of Retained Reads: \" + str(retained_values_counter))\n",
    "    output1.write(\"\\n\" + \"Number of Discarded Reads: \" + str(discarded_values_counter))\n",
    "    output1.write(\"\\n\" + \"Total Number of Reads: \" + str(total_counter))\n",
    "    output1.write(\"\\n\" + \"Percentage of Reads Index Swapped: \" + str((idx_op_counter/total_counter)*100))\n",
    "    output1.write(\"\\n\" + \"Percentage of Expected Paired Index Reads: \" + str((matched_counter/total_counter)*100))\n",
    "    output1.write(\"\\n\")\n",
    "    output1.write(\"min_qscore: \" + min_qscore + \"\\n\")\n",
    "    output1.write(\"\\n\")\n",
    "    output1.write(\"Dual Indexed Pairs\" + \"\\t\" + \"Counts\" + \"\\n\")\n",
    "    \n",
    "    for i in index_values:\n",
    "        output1.write(str(i) + \"\\t\" + str(index_values[i]) + \"\\n\")\n",
    "        j+=1       \n",
    "    output1.write(\"\\n\")\n",
    "    output1.write((\"Index Hopped Pairs\" + \"\\t\" + \"Counts\" + \"\\n\"))\n",
    "    for i in un_matched_index:\n",
    "        output1.write(str(i) + \"\\t\" + str(un_matched_index[i]) + \"\\n\")\n",
    "        m+=1\n",
    "    output1.write(\"\\n\")\n",
    "    output1.write((\"N Indexed Pairs\" + \"\\t\" + \"Counts\" + \"\\n\"))\n",
    "    for i in N_Idx_Dic:\n",
    "        output1.write(str(i) + \"\\t\" + str(N_Idx_Dic[i]) + \"\\n\")\n",
    "        k+=1       \n",
    "   "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
